design/ceph/object/store.md

```diff
@@ -323,31 +323,118 @@ spec:
 
 ```yaml
 spec:
   hosting:
+    advertiseEndpoint:
+      dnsName: my-internal-wildcard-service.mydomain.com
+      port: 8443
+      useTLS: true
     dnsNames:
-    - "my-ingress.mydomain.com"
+      - my-external-ingress.mydomain.com
 ```
 
-A list of hostnames to use for accessing the bucket directly like a subdomain in the endpoint. For example if the ingress service endpoint `my-ingress.mydomain.com` added and the object store contains a bucket named `sample`, then the s3 [virtual host style](https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html) would be `http://sample.my-ingress.mydomain.com`. More details about the feature can be found in [Ceph Documentation](https://docs.ceph.com/en/latest/radosgw/s3/commons/#bucket-and-host-name).
-
-When this feature is enabled the endpoint in OBCs and COSI Bucket Access can be clubbed with bucket name and host name. For OBC  the `BUCKET_NAME` and the `BUCKET_HOST` from the config map combine to `http://$BUCKETNAME.$BUCKETHOST`. For COSI Bucket Access the `bucketName` and the `endpoint` in the `BucketInfo` to `http://bucketName.endpoint`.
+Each `rgw_dns_name` entry must be a valid RFC-1123 hostname, and Rook Operator will perform input validation using k8s apimachinery `IsDNS1123Subdomain()`.
+
+When `rgw_dns_name` is changed for an RGW cluster, all RGWs need to be restarted. To enforce this in Rook, we can apply the `--rgw-dns-name` flag, which will restart RGWs with no user action needed.
+
+This is different from CephObjectZone `customEndpoints` which is used for configuring the object store to replicate and sync data amongst Multisite.
+
+When endpoints are configured in `rgw_dns_names`, the RGW will reject any incoming connections
+intended for endpoints not in the list. Therefore, all endpoints that might be used must be added.
+
+For convenience, and to ensure other CephObjectStore configurations are not rendered unusable when
+users add additional endpoints to `dnsNames`, the following should be added to the `rgw_dns_names`
+list automatically by Rook:
+- the `advertiseEndpoint.dnsName`
+- the default service endpoint for the object store (e.g., `rook-ceph-rgw-my-store.rook-ceph.svc`)
+- CephObjectZone `customEndpoints`
+
+When Rook builds the `rgw_dns_names` list internally, Rook should remove any duplicate entries.
+While Rook add endpoints to the list for safety and convenience, users might add the same endpoints,
+which Rook should not treat as a configuration bug. Rook should also ensure the list ordering is
+consistent between reconciles.
+
+For a `hosting` configuration that includes only `dnsNames`, there is potential internal/developer
+and user confusion when adding more than one entry to the `dnsNames` list. From the user
+perspective, Rook might make undesirable assumptions about which endpoint to advertise for
+CephObjectStoreUser, OBC, or COSI resources. From the internal/developer perspective, it might be
```

Comment:
@travisn I'm reconsidering whether it's really best to make `advertiseEndpoint` required. 

If users don't set this, Rook can continue using the default behavior that it uses when `dnsNames` isn't specified and just return the service endpoint, prioritizing HTTPS when available. That is easy to document, and it means that users who are okay with that don't need to look up what the service endpoint is.

WDYT?

Reply:
Agreed, that sounds good to have the default use the service endpoint when that value is not set.